src/engine/routing_algorithms/alternative_path_mld.cpp

```diff
@@ -0,0 +1,786 @@
+        std::size_t forward_min_index = via_at;
+        std::size_t reverse_min_index = via_at;
+
+        for (std::size_t i = via_at; i > 0; --i)
+        {
+            const auto node = path_nodes[i];
+            const auto weight = forward_heap.GetKey(node);
+
+            if (weight < forward_min_weight)
+            {
+                forward_min_index = i;
+                break;
+            }
+        }
+
+        for (std::size_t i = via_at; i < path_nodes.size(); ++i)
+        {
+            const auto node = path_nodes[i];
+            const auto weight = reverse_heap.GetKey(node);
+
+            if (weight < reverse_min_weight)
+            {
+                reverse_min_index = i;
+                break;
+            }
+        }
+
+        // We now have the ranges [f, via] and [via, l] which we have to test for optimality.
+        // We can do this by checking the node predecessors in the heaps, these are optimal.
+
+        const auto forward_subpath_first = begin(path_nodes) + forward_min_index;
+        const auto forward_subpath_last = begin(path_nodes) + via_at;
+
+        const auto reverse_subpath_first = begin(path_nodes) + via_at;
+        const auto reverse_subpath_last = begin(path_nodes) + reverse_min_index;
+        */
+
+        // Plateaux iff via == parent_in_reverse_search(parent_in_forward_search(via))
+        //
+        // Forward search starts from s, reverse search starts from t:
+        //  - parent_in_forward_search(via) = a
+        //  - parent_in_reverse_search(a) = b  != via and therefore not local optimal
+        //
+        //        via
+        //      .'   '.
+        // s - a - - - b - t
+        //
+        // Care needs to be taken for the edge case where the via node is on the border
+        // of the search spaces and therefore parent pointers may not be valid in heaps.
+        // In these cases we know we can't have local optimality around the via already.
+
+        const auto parent_in_forward = forward_heap.GetData(via).parent;
+        const auto parent_in_reverse = reverse_heap.GetData(via).parent;
+
+        // Edge case where via is at the border of the search spaces;
+        // we may not have seen the node's parent in the dual heap.
+        if (!forward_heap.WasInserted(parent_in_reverse))
+            return true;
+        if (!reverse_heap.WasInserted(parent_in_forward))
+            return true;
+
+        const auto parent_parent_in_forward = forward_heap.GetData(parent_in_reverse).parent;
+        const auto parent_parent_in_reverse = reverse_heap.GetData(parent_in_forward).parent;
+
+        const auto plateaux = parent_parent_in_forward == via && parent_parent_in_reverse == via;
+
+        return !plateaux;
+    };
+
+    return std::remove_if(first, last, is_not_locally_optimal);
+}
+
+// The unpacked paths' sharing in [0, 1] for no sharing and equality, respectively.
+inline double normalizedUnpackedPathSharing(const UnpackedNodes &lhs, const UnpackedNodes &rhs)
```

Comment:
As you already said to me, all of these should probably use weights, normalised to not include access penalties